cis194/week7/haotian/week7/src/JoinList.hs

```diff
@@ -66,3 +67,16 @@ dropJ n l@(Append m l1 l2)
     sizeM = getSize . size
     sizeL = getSize . size . tag 
 dropJ _ _ = Empty
+
+--task 2 takeJ
+takeJ :: (Sized b, Monoid b) => Int -> JoinList b a -> JoinList b a
+takeJ n l1@(Single _ _)
+  | n > 0 = l1
+takeJ n l@(Append m l1 l2)
+  | n >= sizeM m = l
+  | n >= sizeL l1 = l1 +++ takeJ (n - sizeL l1) l2
+  | n > 0 = takeJ n l1
+  where
+    sizeM = getSize . size
+    sizeL = getSize . size . tag 
```

Comment:
FYI - you may shrink further (not tested). 
`Sized b => JoinList b a` is an instance of `Sized`. 

```haskell
takeJ n l@(Append m l1 l2)
  | n >= sizeOf m = l
  | n >= sizeL1 = l1 +++ takeJ (n - sizeL1) l2
  | n > 0 = takeJ n l1
  where
    sizeOf = getSize . size
    sizeL1 = sizeOf l1
```

Reply:
@sinogermany 

I think still need
```
where
    sizeM = getSize . size
    sizeL = sizeM . tag $ l1 
```